### PR TITLE
Minimize mod req

### PR DESCRIPTION
**Pull request by tech4GT**
@jywarren Have a look at the invert module file, that's what the new minimized module requirement will look like!!

### CODE REVIEW

**Comment by jywarren** [comment]
This is really great. I guess i imagine even more could be reduced or automated (as over-ridable default)! Are you thinking of building incrementally? Should we merge this first, refactor all the modules, and then try to simplify even more, so we don't change too much at once?

**Comment by tech4GT** [comment]
@jywarren Actually I was thinking about this and I think we should settle for this right now and we can sort of collect ideas about this over time and once we have some we can get to this again, plus changing the module requirements to this also shortens things out a lot I guess

**Comment by jywarren** [comment]
OK does this break any previous modules or anything? Or is it a "new
feature" that we can do a v0.x.0 bump for? Do you want to write a "before"
vs "after" example of module requirements in the README?

**Comment by tech4GT** [comment]
@jywarren this is not ready right now!! readme and other modules have to be updated and also this should not technically break any previous modules but those modules would contain redundant code, I would have to try it ti be sure. I opened this to get your approval for the module format basically

**Comment by jywarren** [comment]
cool! Yes, i like doing this incrementally too, so let's plan to do this
thoroughly then try getting it even simpler. Thanks!

**Comment by tech4GT** [comment]
@jywarren I am now working on this!! Refactoring modules now!! Also updating readme

**Comment by tech4GT** [comment]
@jywarren I think this is now ready to merge!!

File: src/modules/Crop/Module.js [hunk 1]
```diff
@@ -15,12 +15,7 @@
...
 
-  // TODO: we could also set this to {} if nil in AddModule.js to avoid this line:
-  options = options || {};
-
-  // Tell the UI that a step has been added
-  UI.onSetup(options.step); // we should get UI to return the image thumbnail so we can attach our own UI extensions
-
+  // we should get UI to return the image thumbnail so we can attach our own UI extensions
   // add our custom in-module html ui:
   if (options.step.inBrowser) var ui = require('./Ui.js')(options.step, UI);
```
Comment (jywarren) [inline src/modules/Crop/Module.js]: So if we wanted to customize the UI, this is a good example module. Maybe we should link into this from the README as an example of a customized UI?

**Comment by jywarren** [comment]
Can you put this up at your own fork's `gh-pages` branch so we can test it out? it's a big change!

**Comment by tech4GT** [comment]
@jywarren Sure, doing it now!

**Comment by tech4GT** [comment]
@jywarren I forgot, I have a custom user domain setup, so It'll not work with my fork!!

**Comment by tech4GT** [comment]
Maybe I can host it on some other service?

**Comment by jywarren** [comment]
Want to make a testing GitHub account?

**Comment by tech4GT** [comment]
We can do that

**Comment by tech4GT** [comment]
@jywarren or maybe we can test it out on localhost only for now, will that be possible for you??

**Comment by tech4GT** [comment]
@jywarren I am also thinking maybe after the summer we can setup nightwatch on the project?

**Comment by jywarren** [comment]
I can but it takes more time and I'm traveling today -- I think it's going
to be worthwhile to have the ability to post a demo and will speed up our
workflow for the rest of the summer! Can you figure out a way?

On Jun 12, 2018 11:11 AM, "Varun Gupta" <notifications@github.com> wrote:

@jywarren <https://github.com/jywarren> I am also thinking maybe after the
summer we can setup nightwatch on the project?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/publiclab/image-sequencer/pull/289#issuecomment-396646865>,
or mute the thread

**Comment by tech4GT** [comment]
@jywarren Sure I'll find a way, Safe travels!! :)

**Comment by tech4GT** [comment]
btw I am working on the step API functions today and I wanted to ask about the listeners? we already have onDraw and onComplete in the UI object right, do you want me to make them available on this reference?

**Comment by tech4GT** [comment]
http://tech4gt.me/image-sequencer/examples/ @jywarren try this

**Comment by jywarren** [comment]
This looks good! Do you think this is ready to merge, then? Does it break any previous API or just add features (so `vx.0.0` or `v0.x.0`) version bump?

Could we use the Crop module in some documentation to demonstrate how you can still create a custom UI?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -35,23 +35,17 @@ Any module must follow this basic format:
 ```js
 module.exports = function ModuleName(options,UI) {
 
-  options = options || {};
```
Comment (jywarren) [inline CONTRIBUTING.md]: Here let's say `Module requirements have simplified in version XXX` - see here for previous (and link to this section of the readme from a specific release -- like https://github.com/publiclab/image-sequencer/blob/master/CONTRIBUTING.md#contributing-modules

**Comment by jywarren** [comment]
Hmm, also  i tried the crop drag UI in http://tech4gt.me/image-sequencer/examples/#steps=invert(),colormap(colormap:default),crop(x:0|y:0|w:(100%25)|h:(100%25)) but it doesn't seem to work -- can you take a look?

**Comment by tech4GT** [comment]
@jywarren Oh it's just node module issue!! the image-area-select is not installed I think!! I tried it on localhost though it works fine!!

File: CONTRIBUTING.md (hunk 2 above)
Comment (tech4GT) [reply CONTRIBUTING.md]: @jywarren Got it!! Also since the UI object is still being passed to the module, it can technically do anything with it!!

**Comment by jywarren** [comment]
ah ok can you push up the node modules there too so I can do a final check? Thanks!

**Comment by tech4GT** [comment]
@jywarren done!!

**Comment by jywarren** [comment]
OK, it works really well. Great work. Did you want to bump the version number now?

**Comment by tech4GT** [comment]
@jywarren Thanks, I say lets merge in steps api as well and then bump a major version, meta modules won't break any old functions so that should be a minor update, what say?

**Comment by jywarren** [comment]
Fantastic

**Pull request merged**